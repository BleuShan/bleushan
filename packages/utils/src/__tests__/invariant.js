import invariantWhenInvokedWith from '../__fixtures__/invariant.js'

describe('invariant', () => {
  describe('when invoked', () => {
    describe.each([
      [undefined],
      [{message: 1}],
      [{message: 'string', condition: () => 'ste'}],
      [{message: 'string', condition: ''}],
      [{message: 'string', condition: false, errorType: 'stes'}],
      [{message: 'string', condition: true, errorType: 'stes'}]
    ])('with an invalid configuration, like %o,', (configuration) => {
      it('should throw with the appropriate error', () => {
        expect(invariantWhenInvokedWith(configuration)).toThrowErrorMatchingSnapshot()
      })
    })

    describe('with a valid configuration', () => {
      let configuration
      const conditionMock = jest.fn()

      class CustomError extends Error {
        constructor(message) {
          super(`CustomError: ${message}`)
          Reflect.setPrototypeOf(this, new.target.prototype)
        }
      }

      class CustomErrorWithArgs extends Error {
        constructor(message, ...args) {
          super(message + args.join(','))
          Reflect.setPrototypeOf(this, new.target.prototype)
        }
      }

      describe.each`
        message            | errorType              | args
        ${'default error'} | ${undefined}           | ${undefined}
        ${'some message'}  | ${CustomError}         | ${[]}
        ${'Counting:'}     | ${CustomErrorWithArgs} | ${[1, 2, 3]}
      `('with message: $message and errorType: $errorType', ({message, errorType, args}) => {
        beforeEach(() => {
          configuration = {message, errorType, args}
        })

        describe.each`
          description   | accessType
          ${'a getter'} | ${'get'}
          ${'a method'} | ${undefined}
        `('when the condition is evaluated via $description', ({accessType}) => {
          beforeEach(() => {
            Reflect.defineProperty(configuration, 'condition', {
              configurable: true,
              ...(accessType ? {get: conditionMock} : {value: conditionMock})
            })
          })

          describe.each([[true], [false]])('and it evaluates to %o', (value) => {
            beforeEach(() => {
              conditionMock.mockReturnValue(value)
            })

            if (value) {
              it('should not throw', () => {
                expect(invariantWhenInvokedWith(configuration)).not.toThrow()
              })
            } else {
              it('should throw with the appropriate error', () => {
                expect(invariantWhenInvokedWith(configuration)).toThrowErrorMatchingSnapshot()
              })
            }
          })
        })
      })
    })
  })
})
