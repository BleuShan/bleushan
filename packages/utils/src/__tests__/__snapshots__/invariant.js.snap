// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invariant when invoked with a valid configuration with message: Counting: and errorType: [Function CustomErrorWithArgs] when the condition is evaluated via a getter and it evaluates to false should throw with the appropriate error 1`] = `"Counting:1,2,3"`;

exports[`invariant when invoked with a valid configuration with message: Counting: and errorType: [Function CustomErrorWithArgs] when the condition is evaluated via a method and it evaluates to false should throw with the appropriate error 1`] = `"Counting:1,2,3"`;

exports[`invariant when invoked with a valid configuration with message: default error and errorType: undefined when the condition is evaluated via a getter and it evaluates to false should throw with the appropriate error 1`] = `"default error"`;

exports[`invariant when invoked with a valid configuration with message: default error and errorType: undefined when the condition is evaluated via a method and it evaluates to false should throw with the appropriate error 1`] = `"default error"`;

exports[`invariant when invoked with a valid configuration with message: some message and errorType: [Function CustomError] when the condition is evaluated via a getter and it evaluates to false should throw with the appropriate error 1`] = `"CustomError: some message"`;

exports[`invariant when invoked with a valid configuration with message: some message and errorType: [Function CustomError] when the condition is evaluated via a method and it evaluates to false should throw with the appropriate error 1`] = `"CustomError: some message"`;

exports[`invariant when invoked with an invalid configuration, like {
  message: 'string',
  condition: [Function: bound condition] { [length]: 0, [name]: 'bound condition' }
}, should throw with the appropriate error 1`] = `"expected condition be able to be evaluated to a boolean, received: ste"`;

exports[`invariant when invoked with an invalid configuration, like { message: 'string', condition: '' }, should throw with the appropriate error 1`] = `"expected condition be able to be evaluated to a boolean, received: "`;

exports[`invariant when invoked with an invalid configuration, like { message: 'string', condition: false, errorType: 'stes' }, should throw with the appropriate error 1`] = `"errorContructor is not a function"`;

exports[`invariant when invoked with an invalid configuration, like { message: 'string', condition: true, errorType: 'stes' }, should throw with the appropriate error 1`] = `"errorContructor is not a function"`;

exports[`invariant when invoked with an invalid configuration, like { message: 1 }, should throw with the appropriate error 1`] = `"expected message to be a string, received: 1"`;
